File: HW2/invest.py
""" Homework 2
Computational Frameworks, POLS 398
Spring 2012, Duke University
Matt Dickenson """
# try using Exceptions

import random
import ystockquote

class Portfolio(object):

     def __init__(self):
        self.transactions = ""
        self.cashBalance = 0 
        self.stockBalance = {} # this will be a dict of dicts
        self.mfBalance = {}

     def __str__(self):
        self.stock_list = ""
        for i in range(len(self.stockBalance)):
            new_string = str(self.stockBalance[i]["shares"]) + ' ' + str(self.stockBalance[i]["symbol"])
            if len(self.stockBalance) > i : # may need to be i + 1 
                new_string += '\n'
            self.stock_list += new_string

        self.mf_list = ""
        for i in range(len(self.mfBalance)):
            new_string = str(self.mfBalance[i]["shares"]) + ' ' + str(self.mfBalance[i]["symbol"])
            if len(self.mfBalance) > i : 
                new_string += '\n'
            self.mf_list += new_string
        
        return "cash: $%s \nstock: %s \nmutual funds: %s" % (self.cashBalance, self.stock_list, self.mf_list)

     def history(self):
        print self.transactions
     
     def addCash(self, amount):
        self.cashBalance += amount
        self.transactions += ("$%.2f added.\n" % amount)

     def withdrawCash(self, amount):
        self.cashBalance -= amount # don't withdraw negative money
                                   # check the amt to be withdrawn
        self.transactions += ("$%.2f withdrawn. \n" % amount)

     def buyStock(self, shares, stock): # what if i'm buying for the 2
        purchase_amount = shares * stock.price
        if (shares % 1 == 0) & (purchase_amount <= self.cashBalance):
            self.cashBalance -= purchase_amount
           self.stockBalance[len(self.stockBalance)] = {"symbol": stock.name, "shares": shares, "object": stock} # the way this is being used it might as well be a list
                                                                                                                 # what is returned if this works?
              
            self.transactions += ("Purchased %s shares of %s stock at $%.2f per share, for a total of $%.2f.\n" % (shares, stock.name, stock.price, purchase_amount))

        elif (shares % 1 == 0) & (purchase_amount > self.cashBalance):
            return "That purchase exceeds your cash balance of %.2f. Please enter a smaller purchase." % self.cashBalance

        elif (shares % 1 != 0):
            return "Fractional shares cannot be purchased. Please enter a whole number of shares."

        else: return "I have no idea what you mean. Please try again."

     def sellStock(self, symbol, share_amt):
        for i in  range(len(self.stockBalance)):
            if self.stockBalance[i]['symbol'] == symbol: 
                num_shares_owned = self.stockBalance[i]['shares']
                stock_to_sell = self.stockBalance[i]['object']
                x = i
        if num_shares_owned >= share_amt:
            stock_to_sell.calc_sell_price()
            total_sell_price = stock_to_sell.sell_price * share_amt
            self.cashBalance += total_sell_price
            self.transactions += "Sold %s shares of %s at a price of $%.2f, for a total of $%.2f.\n" % (share_amt, symbol, stock_to_sell.sell_price, total_sell_price)
            self.stockBalance[x]['shares'] = 0 # this should be num_shares_owned - share_amt
                                               # how do I know if it worked?
 
     def buyMutualFund(self, shares, fund):
        purchase_amount = shares * fund.price
        if (purchase_amount <= self.cashBalance):
            self.cashBalance -= purchase_amount
            self.mfBalance[len(self.mfBalance)] = {"symbol": fund.name, "shares": shares, "object": fund}
            
            self.transactions += ("Purchased %s shares of %s fund at $%.2f per share, for a total of $%.2f.\n" % (shares, fund.name, fund.price, purchase_amount))

        elif (purchase_amount > self.cashBalance):
            return "That purchase exceeds your cash balance of %s. Please enter a smaller purchase." % self.cashBalance

        else: return "I have no idea what you mean. Please try again."

     def sellMutualFund(self, symbol, share_amt):
        for i in  range(len(self.mfBalance)):
            if self.mfBalance[i]['symbol'] == symbol: 
                num_shares_owned = self.mfBalance[i]['shares']
                fund_to_sell = self.mfBalance[i]['object']
                x = i
        if num_shares_owned >= share_amt:
            fund_to_sell.calc_sell_price()
            total_sell_price = fund_to_sell.sell_price * share_amt
            self.cashBalance += total_sell_price
            self.transactions += "Sold %s shares of %s at a price of $%2f, for a total of $%.2f.\n" % (share_amt, symbol, fund_to_sell.sell_price, total_sell_price)
            self.mfBalance[x]['shares'] = 0
            
        else:
            return "You do not have that many shares." 


class Stock(object):
    def __init__(self, price, name):
        self.price = price
        self.name = name
        # make object printable with a def __str__()--see Matthias' code
 
    def calc_sell_price(self):
        self.sell_price = self.price * random.uniform(0.5, 1.5)
        self.sell_price = round(self.sell_price, 2)

class realStock(object):
    def __init__(self, symbol):
        self.price = float(ystockquote.get_price(symbol))
        self.name = symbol
        self.market_cap = ystockquote.get_market_cap(symbol)
        self.ebitda = ystockquote.get_ebitda(symbol)
        self.price_52_week_high = ystockquote.get_52_week_high(symbol)
        self.price_52_week_low = ystockquote.get_52_week_low(symbol)
        self.PE_ratio = ystockquote.get_price_earnings_ratio(symbol)

class MutualFund(object):

    def __init__(self, name):
        self.name = name
        self.price = 1
        
    def calc_sell_price(self):
        self.sell_price = self.price * random.uniform(0.9, 1.2)
        self.sell_price = round(self.sell_price, 2)
        

File: HW2/invest.rb
# HW2, PS398, implementation in Ruby

# First we add a round_to method for Float numbers
class Float
  def round_to(x)
    (self * 10**x).round.to_f / 10**x
  end

  def ceil_to(x)
    (self * 10**x).ceil.to_f / 10**x
  end

  def floor_to(x)
    (self * 10**x).floor.to_f / 10**x
  end
end


class Portfolio < Object

  attr_reader :transactions, :cashBalance, :stockBalance, :mfBalance

  def initialize
    @transactions = ''
    @cashBalance = 0
    @stockBalance = Hash.new
    @mfBalance = Hash.new
  end

  def print
    @stock_list = "" # refresh lists
    @mf_list = ""

    temp_long_str = ''
    @stockBalance.each_key do |k|
      temp_str = String(k) + ' ' + String(@stockBalance[k]["shares"]) + "\n"
      temp_long_str += temp_str
    end
    @stock_list = temp_long_str[0..-2] # removes last newline

    temp_long_str = ''
    @mfBalance.each_key do |k|
      temp_str = String(k) + ' ' + String(@mfBalance[k]["shares"]) + "\n" 
      # TODO: round to 2 digits if numshares is float
      temp_long_str += temp_str
    end
    @mf_list = temp_long_str[0..-2] # removes last newline

    puts "Cash: %.2f \nStock: #{@stock_list} \nMutual Funds: #{@mf_list}" % @cashBalance
    puts
  end

  def history
    puts self.transactions
  end

  def addCash(amount)
    @cashBalance += amount
    @transactions += ("$%.2f added.\n" % amount)
  end

  def withdrawCash(amount)
    if amount < 0
      puts "Negative amounts not allowed."
      return
    end
    @cashBalance -= amount
    @transactions += ("$%.2f withdrawn.\n" % amount)
  end

  def buyStock(shares, stock)
    purchase_amount = shares * stock.price
    if (shares % 1 == 0) & (purchase_amount <= @cashBalance)
      @cashBalance -= purchase_amount
      @stockBalance[stock.name] = {"shares" => shares, "object" => stock}
      @transactions += ("Purchased #{shares} shares of #{stock.name} stock at $#{stock.price} per share, for a total of $#{purchase_amount}\n")
      puts "#{stock.name} successfully purchased."
    elsif (shares % 1 == 0) & (purchase_amount  >@cashBalance)
      puts "That purchase exceeds your cash balance of $#{cash_balance}."
    elsif (shares % 1 != 0)
      puts "Fractional shares cannot be purchased."
    end
  end

  def buyMutualFund(shares, fund)
    purchase_amount = shares * fund.price
    if (purchase_amount <= @cashBalance)
      @cashBalance -= purchase_amount
      @mfBalance[fund.name] = {'shares' => shares, 'object' => fund}
      @transactions += ("Purchased #{shares} shares of #{fund.name} fund at $#{fund.price} for a total of #{purchase_amount}.\n")
    else
      puts "That purchase exceeds your cash balance of #{@cashBalance}."
    end
  end

  def sellStock(symbol, shares_to_sell)
    #puts @stockBalance[symbol]
    shares_owned = @stockBalance[symbol]["shares"]
    if shares_owned >= shares_to_sell
      stock_obj = @stockBalance[symbol]["object"]
      stock_obj.calc_sell_price
      sell_price = stock_obj.sell_price
      total_sell_price = sell_price * shares_to_sell
      @cashBalance += total_sell_price
      temp_txt = "Sold #{ shares_to_sell } shares of #{symbol} at a price of $#{sell_price} for a total of $#{total_sell_price}."
      @transactions += temp_txt + "\n"
      puts temp_txt
      @stockBalance[symbol]["shares"] -= shares_to_sell
    else
      puts "You do not have that many shares."
    end
  end

  def sellMutualFund(symbol, shares_to_sell)
    shares_owned = @mfBalance[symbol]["shares"]
    if shares_owned >= shares_to_sell
      mf_obj = @mfBalance[symbol]["object"]
      mf_obj.calc_sell_price
      sell_price = mf_obj.sell_price
      total_sell_price = (sell_price * shares_to_sell).round_to(2)
      @cashBalance += total_sell_price
      temp_txt = "Sold #{ shares_to_sell } shares of #{symbol} at a price of $#{sell_price} for a total of $#{total_sell_price}."
      @transactions += temp_txt + "\n"
      puts temp_txt
      @mfBalance[symbol]["shares"] -= shares_to_sell
      #@mfBalance[symbol]["shares"].round_to(2)
    else
      puts "You do not have that many shares."
    end
  end

end


class Stock < Object

  attr_accessor :price, :name 
  attr_reader :sell_price

  def initialize(price, name)
    @price = price
    @name = name
  end

  def print
    puts "Name: #{@name} \nPrice: #{@price}"
    puts
  end

  def calc_sell_price
    @sell_price = ( @price * (0.5+rand)).round_to(2)
  end
end


class MutualFund < Object

  attr_accessor :price, :name 
  attr_reader :sell_price

  def initialize(name)
    @price = 1
    @name = name
  end

  def print
    puts "Name: #{@name}"
    puts
  end

  def calc_sell_price
    @sell_price = (0.9+rand*0.3).round_to(2)
  end
end


    
# TODO: add exceptions for when people try to sell stocks they don't own
# TODO: make all prints of cashBalance round to 2 decimal places
